statesync/dispatcher.go

```diff
@@ -0,0 +1,322 @@
+func (d *dispatcher) dispatch(peer p2p.NodeID, height int64) (chan *types.LightBlock, error) {
+	d.mtx.Lock()
+	defer d.mtx.Unlock()
+	ch := make(chan *types.LightBlock, 1)
+
+	// check if the dispatcher is running or not
+	if !d.running {
+		close(ch)
+		return ch, errDisconnected
+	}
+
+	// this should happen only if we add the same peer twice (somehow)
+	if _, ok := d.calls[peer]; ok {
+		close(ch)
+		return ch, errPeerAlreadyBusy
+	}
+	d.calls[peer] = ch
+
+	// send request
+	d.requestCh <- p2p.Envelope{
+		To: peer,
+		Message: &ssproto.LightBlockRequest{
+			Height: uint64(height),
+		},
+	}
+	return ch, nil
+}
+
+// release appends the peer back to the list and deletes the allocated call so
+// that a new call can be made to that peer
+func (d *dispatcher) release(peer p2p.NodeID) {
+	d.mtx.Lock()
+	defer d.mtx.Unlock()
+	if call, ok := d.calls[peer]; ok {
+		close(call)
+		delete(d.calls, peer)
+	}
+	d.availablePeers.Append(peer)
+}
+
+//----------------------------------------------------------------
+
+// blockProvider is a p2p based light provider which uses a dispatcher connected
+// to the state sync reactor to serve light blocks to the light client
+//
+// TODO: This should probably be moved over to the light package but as we're
+// not yet officially supporting p2p light clients we'll leave this here for now.
+type blockProvider struct {
+	peer       p2p.NodeID
+	chainID    string
+	timeout    time.Duration
+	dispatcher *dispatcher
+}
+
+func (p *blockProvider) LightBlock(ctx context.Context, height int64) (*types.LightBlock, error) {
+	// FIXME: The provider doesn't know if the dispatcher is still connected to
+	// that peer. If the connection is dropped for whatever reason the
+	// dispatcher needs to be able to relay this back to the provider so it can
+	// return ErrConnectionClosed instead of ErrNoResponse
+	ctx, cancel := context.WithTimeout(ctx, p.timeout)
+	defer cancel()
+	lb, _ := p.dispatcher.lightBlock(ctx, height, p.peer)
+	if lb == nil {
+		return nil, provider.ErrNoResponse
+	}
+
+	if err := lb.ValidateBasic(p.chainID); err != nil {
+		return nil, provider.ErrBadLightBlock{Reason: err}
+	}
+
+	return lb, nil
+}
+
+// ReportEvidence should allow for the light client to report any light client
+// attacks. This is a no op as there currently isn't a way to wire this up to
+// the evidence reactor (we should endeavor to do this in the future but for now
+// it's not critical for backwards verification)
+func (p *blockProvider) ReportEvidence(ctx context.Context, ev types.Evidence) error {
+	return nil
+}
+
+// String implements stringer interface
+func (p *blockProvider) String() string { return string(p.peer) }
+
+//----------------------------------------------------------------
+
+// peerList is a rolling list of peers. This is used to distribute the load of
+// retrieving blocks over all the peers the reactor is connected to
+type peerlist struct {
```

Comment:
unless I'm missing something, it'd probably be better to use a map or something to randomize the order? I think FIFO is likely a sub par strategy, and makes the data structure less effective.

Reply:
Can you elaborate on why you think a FIFO strategy is inefficient? Are you thinking that if two peers have the same peer list and both start state sync at the same time they will be pinging the same peer at the same time?